prov/gni/src/gnix_msg.c

```diff
@@ -1458,47 +2248,106 @@ static int _gnix_send_req(void *arg)
-
-		tag = GNIX_SMSG_T_RNDZV_START;
-		tdesc->rndzv_start_hdr.flags = req->msg.send_flags;
-		tdesc->rndzv_start_hdr.imm = req->msg.imm;
-		tdesc->rndzv_start_hdr.msg_tag = req->msg.tag;
-		tdesc->rndzv_start_hdr.mdh = req->msg.send_md->mem_hndl;
-		tdesc->rndzv_start_hdr.addr = req->msg.send_addr;
-		tdesc->rndzv_start_hdr.len = req->msg.send_len;
-		tdesc->rndzv_start_hdr.req_addr = (uint64_t)req;
-
-		if (req->msg.send_addr & GNI_READ_ALIGN_MASK) {
-			tdesc->rndzv_start_hdr.head =
-				*(uint32_t *)(req->msg.send_addr &
-					      ~GNI_READ_ALIGN_MASK);
-			GNIX_INFO(FI_LOG_EP_DATA,
-				  "Sending %d unaligned head bytes (%x)\n",
-				  GNI_READ_ALIGN -
-				  (req->msg.send_addr & GNI_READ_ALIGN_MASK),
-				  tdesc->rndzv_start_hdr.head);
-		}
+		if (req->msg.send_flags & GNIX_MSG_IOV) {
+			assert(req->msg.send_iov_md);
+			tag = GNIX_SMSG_T_RNDZV_IOV_START;
+			tdesc->rndzv_iov_start_hdr.flags = req->msg.send_flags;
+			tdesc->rndzv_iov_start_hdr.imm = req->msg.imm;
+			tdesc->rndzv_iov_start_hdr.msg_tag = req->msg.tag;
+			tdesc->rndzv_iov_start_hdr.iov_cnt =
+				req->msg.send_iov_cnt;
+			tdesc->rndzv_iov_start_hdr.req_addr = (uint64_t) req;
+			tdesc->rndzv_iov_start_hdr.send_addr =
+				req->msg.send_addr;
+			tdesc->rndzv_iov_start_hdr.send_len = req->msg.send_len;
+			data_len = sizeof(struct iovec) * req->msg.send_iov_cnt;
+			data = (void *) req->msg.send_addr;
+
+#if ENABLE_DEBUG
+			for (i = 0; i < req->msg.send_iov_cnt; i++) {
+				GNIX_DEBUG(FI_LOG_EP_DATA, "iov_base[%d] = %p, "
+					   "iov_len[%d] = %lu\n", i,
+					   ((struct iovec *)data)[i].iov_base,
+					   i,
+					   ((struct iovec *)data)[i].iov_len);
+			}
+#endif
 
-		if ((req->msg.send_addr + req->msg.send_len) &
-		    GNI_READ_ALIGN_MASK) {
-			tdesc->rndzv_start_hdr.tail =
-				*(uint32_t *)((req->msg.send_addr +
-					       req->msg.send_len) &
-					      ~GNI_READ_ALIGN_MASK);
-			GNIX_INFO(FI_LOG_EP_DATA,
-				  "Sending %d unaligned tail bytes (%x)\n",
-				  (req->msg.send_addr + req->msg.send_len) &
-				  GNI_READ_ALIGN_MASK,
-				  tdesc->rndzv_start_hdr.tail);
-		}
+			/*
+			 * Allocate the header and place the memory descriptors
+			 * in the header.
+			 * TODO: Is there a better way to pack the mds?
+			 * hdr_format: [iov_start_hdr|mdh_0mdh_1mdh_2...]
+			 */
+			hdr_len = sizeof(tdesc->rndzv_iov_start_hdr) +
+				(sizeof(gni_mem_handle_t) *
+				 tdesc->rndzv_iov_start_hdr.iov_cnt);
+
+			hdr = malloc(hdr_len);
+			assert(hdr != NULL);
+
+			/* Record this as it must be free'd upon completion */
+			req->msg.smsg_iov_hdr_addr = (uint64_t) hdr;
+
+			memcpy(hdr, &tdesc->rndzv_iov_start_hdr,
+			       sizeof(tdesc->rndzv_iov_start_hdr));
+
+			GNIX_DEBUG(FI_LOG_EP_DATA, "Packing header..."
+				   "req_addr = %p, req->msg.send_len = %lu\n",
+				   ((struct gnix_smsg_rndzv_iov_start_hdr *)
+				    hdr)->req_addr, req->msg.send_len);
+			/* Pack up the memory descriptors */
+
+			/* There must be a better way.. */
```

Comment:
From another comment, I'm guessing that you'll need to allocate your own buffer to hold iov addr/length information (since the user's iov buffer may not be available when fi_sendv() returns).  If you need to make your own iov buffer, you could size it to hold the MDs as well.  Then you could pass that as the data pointer for your control message (you currently pass in the user's iov buffer as the data).  Then you wouldn't have to reallocate the header here.